util/src/main/java/com/google/cloud/hadoop/util/GrpcErrorTypeExtractor.java

```diff
@@ -34,8 +39,40 @@ public ErrorType getErrorType(Exception error) {
         return ErrorType.NOT_FOUND;
       case OUT_OF_RANGE:
         return ErrorType.OUT_OF_RANGE;
+      case ALREADY_EXISTS:
+        return ErrorType.ALREADY_EXISTS;
+      case FAILED_PRECONDITION:
+        return ErrorType.FAILED_PRECONDITION;
       default:
         return ErrorType.UNKNOWN;
     }
   }
+
+  @Override
+  public boolean bucketAlreadyExists(Exception e) {
+    ErrorType errorType = getErrorType(e);
+    if (errorType == ErrorType.ALREADY_EXISTS) return true;
+    // The gRPC API currently throws a FAILED_PRECONDITION status code instead of ALREADY_EXISTS,
+    // so we handle both these conditions in the interim. This can be removed once the status codes
+    // are fixed.
+    else if (errorType == ErrorType.FAILED_PRECONDITION) {
+      StatusRuntimeException statusRuntimeException = getStatusRuntimeException(e);
+      return statusRuntimeException != null
+          && BUCKET_ALREADY_EXISTS_MESSAGE.equals(statusRuntimeException.getMessage());
+    }
+    return false;
+  }
+
+  /** Extracts StatusRuntimeException from the Exception, if it exists. */
+  @Nullable
+  private StatusRuntimeException getStatusRuntimeException(Exception e) {
+    Throwable cause = e;
+    while (cause != null) {
```

Comment:
Can this go into an infinite loop?

Reply:
Nice catch, Added check for self-reference so it doesnt go into infinite loop.